### PR TITLE
Extending Middleare/Interceptor for distributed tracing

### PR DESCRIPTION
**Pull request by martinostvik**
This pull request enhances our tracing capabilities by extending and replacing the existing gRPC tracing interceptor and Echo middleware. We're incorporating a "plug-and-play" interceptor and middleware solution from the [DataDog dd-trace-go repository](https://github.com/DataDog/dd-trace-go/blob/main/contrib/).

gRPC Interceptor: We extend the functionality provided by the gRPC Streaming and Unary interceptors available in the [contrib/google.golang.org/grpc/server.go](https://github.com/DataDog/dd-trace-go/blob/main/contrib/google.golang.org/grpc/server.go) package. These interceptors come preconfigured with distributed tracing capabilities and can be easily customized to suit our requirements. We're utilizing the default options for now to ensure compatibility with the existing library.

Echo Middleware: Additionally, we're replacing the existing Echo middleware with the one from the [contrib/labstack/echo.v4](https://github.com/DataDog/dd-trace-go/tree/main/contrib/labstack/echo.v4) package in the DataDog dd-trace-go repository. This middleware offers out-of-the-box distributed tracing support and can be configured based on specific needs.

Both the gRPC interceptor and Echo middleware are equipped with ready-to-use distributed tracing capabilities, empowering us to seamlessly trace our requests across different services. Moreover, the gRPC interceptor package also provides a [client interceptor](https://github.com/DataDog/dd-trace-go/blob/main/contrib/google.golang.org/grpc/client.go) that facilitates making gRPC requests with traces seamlessly injected into the context.

The feature is "opt-in/feature flagged" by setting an environment variable by the consumer, please see README or internal/env.go in this PR.

For more context, refer to [this engineering issue](https://github.com/coopnorge/engineering-issues/issues/336).

Closes: #156

### CODE REVIEW

**Comment by nhhagen** [comment]
@martinostvik the functions that we are wrapping in this PR supports the passing of `opts... Option`. Is this something we should support?

We might also want to wait changing the signature of the functions until we have figured out how the API for the Go module should evolve.

**Comment by bendiknesbo** [comment]
>We might also want to wait changing the signature of the functions until we have figured out how the API for the Go module should evolve.

I can't see any changing of signatures in this PR, except for the removal of the internal package for holding a span in ctx, which is already handled by DD.

Do note that this PR introduces a difference in the reported data to DD, namely  from `tracer.ResourceName("grpc.request")` to `tracer.ResourceName(info.FullMethod)`, and `operation` changing from `info.FullMethod` to `cfg.spanName`

**Comment by nhhagen** [comment]
@coopnorge/cloud-platform and @bendiknesbo is that something we need to address?

**Comment by bendiknesbo** [comment]
I believe it will be an improvement, as you will be able to differentiate how well your different RPCs are doing, instead of bundling them together in a "grpc.request"-resource.

For reference, the default value of `cfg.spanName` (without sending any opts) will be `grpc.server.request` or `grpc.request`.


Basically, these two values will be swapped:
Screenshot 2023-10-04 at 13 51 22

**Comment by nhhagen** [comment]
Unless anyone objects, I think we should move forward with this change.

**Review by hawi74 — APPROVED** [review]

**Review by nhhagen — DISMISSED** [review]

**Comment by nhhagen** [comment]
@martinostvik can you create a new tag and release once you merge.

**Review by bendiknesbo — DISMISSED** [review]

**Comment by martinostvik** [comment]
> @martinostvik can you create a new tag and release once you merge.

I will make sure to tag the new release. And yes I think it would be smart to test this first.
We could merge this and tag it with a "beta" suffix, so that I can test both he echo middleware and the gRPC interceptor.
Or I can test the contrib version of echo and grpc directly before proceeding with this PR.
Thoughts @nhhagen?

**Comment by bendiknesbo** [comment]
This kind of plays into the issues of how we version our libraries.
Where I'm coming from, we would create git-tags for all commits (that pass CI), even in branches. For branches, the version would be like `v1.2.3.-rc.1`, with the number after `rc.` increasing for each RC, until a new non-RC version was created, at which point, the RC-version would start at 1 again.
This should be enough to allow you to consume this lib from your go-app, but we probably need some gates that prevent/discourage us from deploying apps with RC-dependencies to production.

**Comment by nhhagen** [comment]
So for now we do not need to merge it to test it, just create a tag with an RC version, that will allow you to pull the module.

@bendiknesbo I totally agree, we have issues looking into this, we just need some capacity land and implement the initial version of the strategy.

See https://github.com/coopnorge/engineering-issues/issues/325 for more details.

**Comment by martinostvik** [comment]
> So for now we do not need to merge it to test it, just create a tag with an RC version, that will allow you to pull the module.
> 
> @bendiknesbo I totally agree, we have issues looking into this, we just need some capacity land and implement the initial version of the strategy.
> 
> See [coopnorge/engineering-issues#325](https://github.com/coopnorge/engineering-issues/issues/325) for more details.

testing gRPC interceptor here: https://github.com/coopnorge/store-information-service/pull/388

**Comment by nhhagen** [comment]
To ensure a clean rollout of this feature, I think we might want to hide it behind a feature flag or configuration option so that we can configure this per environment. When we are happy with how it works in a few services we remove the feature flag and make distributed tracing the default/only behaviour.

We should still validate that code works before using it in production systems.

**Comment by martinostvik** [comment]
> To ensure a clean rollout of this feature, I think we might want to hide it behind a feature flag or configuration option so that we can configure this per environment. When we are happy with how it works in a few services we remove the feature flag and make distributed tracing the default/only behaviour.
> 
> We should still validate that code works before using it in production systems.

Is it possible to import and use two different versions of the same go package?

**Comment by bendiknesbo** [comment]
> Is it possible to import and use two different versions of the same go package?

Only if they differ in major versions.

**Comment by nhhagen** [comment]
That leave feature flagging as the only option to "silently" switch this behaviour.

**Comment by martinostvik** [comment]
> That leave feature flagging as the only option to "silently" switch this behaviour.

Not sure I really understand how you want to feature flag it. These are the two options I see.

Option 1: Use `gopkg.in/DataDog/dd-trace-go.v1/contrib/google.golang.org/grpc` directly as you can't switch between `github.com/coopnorge/go-datadog-lib` RC version and main version during runtime without different major version. Downside here is that you are not really testing the actual implmentation in `github.com/coopnorge/go-datadog-lib`

```go
import (
    datadogMiddleware "github.com/coopnorge/go-datadog-lib/v2/middleware/grpc"
    ddGrpc "gopkg.in/DataDog/dd-trace-go.v1/contrib/google.golang.org/grpc"
)

func NewGrpcServer() {
    ...
    interceptor := datadogMiddleware.TraceUnaryServerInterceptor()
    if feature {
        interceptor = dGrpc.ddGrpc.UnaryServerInterceptor()
    }
}

```

Option 2: Feature flag from within the `github.com/coopnorge/go-datadog-lib` package by environment variable

```go
package grpc

import (
    "google.golang.org/grpc"
    ddGrpc "gopkg.in/DataDog/dd-trace-go.v1/contrib/google.golang.org/grpc"
)

// TraceUnaryServerInterceptor for Datadog Log Integration, middleware will create span that can be used from context
func TraceUnaryServerInterceptor() grpc.UnaryServerInterceptor {
    if viper.GetBool("dd_enable_experimental_tracing_interceptor") {
        return experimentalUnaryServerInterceptor()
    }
    return legacyUnaryServerInterceptor()
}

func legacyUnaryServerInterceptor() grpc.UnaryServerInterceptor {
    // existing implementation before this change
}

func experimentalUnaryServerInterceptor() grpc.UnaryServerInterceptor {
    return ddGrpc.UnaryServerInterceptor()
}
```

**Comment by bendiknesbo** [comment]
Option 3:
```go
package grpc

import (
    "google.golang.org/grpc"
    ddGrpc "gopkg.in/DataDog/dd-trace-go.v1/contrib/google.golang.org/grpc"
)

// TraceUnaryServerInterceptor for Datadog Log Integration, middleware will create span that can be used from context
func TraceUnaryServerInterceptor() grpc.UnaryServerInterceptor {
    return legacyUnaryServerInterceptor()
}

// TraceUnaryServerInterceptorExperimental is experimental, and will be removed in the next non-pre-release version. Used for testing a new way of setting up the interceptor.
func TraceUnaryServerInterceptorExperimental() grpc.UnaryServerInterceptor {
    return experimentalUnaryServerInterceptor()
}

func legacyUnaryServerInterceptor() grpc.UnaryServerInterceptor {
    // existing implementation before this change
}

func experimentalUnaryServerInterceptor() grpc.UnaryServerInterceptor {
    return ddGrpc.UnaryServerInterceptor()
}
```

`go-datadog-lib/tracing` also needs to support both versions.

**Comment by nhhagen** [comment]
I would go with Option 2 or Option 3. They have both their pros and their cons.

With Option 2 we can start with making the feature opt-in, then once we want that to be the default we can make it opt-out.

Option 2 also ensure that we do not break the API.

For a pre-release I would go with Option 3.

For a stable release Option 2 might be better, but it is hard to tell without exploring the edge cases.

**Comment by martinostvik** [comment]
> I would go with Option 2 or Option 3. They have both their pros and their cons.
> 
> With Option 2 we can start with making the feature opt-in, then once we want that to be the default we can make it opt-out.
> 
> Option 2 also ensure that we do not break the API.
> 
> For a pre-release I would go with Option 3.
> 
> For a stable release Option 2 might be better, but it is hard to tell without exploring the edge cases.

Let's go for option 2 then, here is a PR for that: https://github.com/coopnorge/go-datadog-lib/pull/158
FYI: It will merge into this branch `distributed_tracing` **not** `main`

File: middleware/echo/echo.go
Comment (arunpoudel) [inline middleware/echo/echo.go:1]: Should we also have this as opt-in rather than replace it? Or is it not worth it?

**Review by arunpoudel — APPROVED** [review]

File: middleware/echo/echo.go
Comment (martinostvik) [reply middleware/echo/echo.go:1]: Yes, this is now updated as a opt-in feature.

**Review by bendiknesbo — APPROVED** [review]

**Review by hawi74 — APPROVED** [review]

**Comment by martinostvik** [comment]
@nhhagen are you OK with this?

**Comment by nhhagen** [comment]
@martinostvik yes, as long as we have tested this in helloworld?

**Comment by martinostvik** [comment]
> @martinostvik yes, as long as we have tested this in helloworld? We should also probably roll this out for helloworld first, then roll it out on the less critical services first.
> 
> This is the first major change to go-datadog-lib in a while.

PR for updating go-datadog-lib to this branch in helloworld: https://github.com/coopnorge/helloworld/pull/481

File: README.md [hunk 1]
```diff
@@ -73,3 +73,20 @@ To preview the documentation:
...
+When this environment variable is set, the package utilizes an experimental
+tracing interceptor from
+[dd-trace-go](https://github.com/DataDog/dd-trace-go/blob/main/contrib/google.golang.org/grpc/server.go)
+for gRPC servers, enhancing distributed tracing capabilities.
+Additionally, for the Echo middleware, the package leverages
+the interceptor from
+[dd-trace-go Echo Middleware](https://github.com/DataDog/dd-trace-go/tree/main/contrib/labstack/echo.v4).
+
+In a future release, these experimental tracing features will become
+the default behavior, eliminating the need for the feature flag.
```
Comment (nhhagen) [inline README.md:92]: This should be moved to `docs/index.md`

Comment (nhhagen) [reply README.md:92]: Moving this in a separate PR

**Pull request merged**